src/platform/packages/shared/kbn-esql-ast/src/pretty_print/basic_pretty_printer.ts

```diff
@@ -162,6 +180,62 @@ export class BasicPrettyPrinter {
     return formatted;
   }
 
+  protected decorateWithSubstitutedParameters<T extends ESQLAstBaseItem>(
+    node: T,
+    formatted: string
+  ): string {
+    if (isColumn(node)) {
+      const argParam = node.args.find((arg) => isParamLiteral(arg));
+      if (argParam) {
+        formatted = this.replaceLiteralWithParameter(argParam, formatted);
+      }
+    } else if (isLiteral(node) && isParamLiteral(node)) {
+      formatted = this.replaceLiteralWithParameter(node, formatted);
+    }
+
+    return formatted;
+  }
+
+  protected replaceLiteralWithParameter(node: ESQLAstExpressionNode, formatted: string): string {
+    if (!isLiteral(node) || !isParamLiteral(node)) {
+      return formatted;
+    }
+
+    const resolveParamValue = (): FieldValue | undefined => {
+      switch (node.paramType) {
+        case 'named':
+        case 'positional':
+          return this.namedParameters[node.value];
+        default:
+          return this.positionalParameters[this.positionalParameterIndex++];
+      }
+    };
+
+    const buildLiteral = (value?: FieldValue) => {
+      if (value === null) {
+        return Builder.expression.literal.nil();
+      }
+
+      switch (typeof value) {
+        case 'number':
+          return Builder.expression.literal.numeric({
+            value,
+            literalType: Number.isInteger(value) ? 'integer' : 'double',
+          });
+        default:
+          return Builder.expression.literal.string(String(value));
+      }
+    };
+
+    const value = resolveParamValue();
+
+    return value === undefined
+      ? formatted
+      : node.paramKind === '?'
+      ? LeafPrinter.literal(buildLiteral(value))
+      : LeafPrinter.identifier(Builder.identifier(String(value)));
+  }
```

Comment:
If we allowed a substitution to be any AST node and a map of substitutions would a single flat map, this could be simplified to:

```suggestion
private printParamWithSubstitution(param: ESQLAstParamLiteral): string {
  const substitution = this.opts.paramSubstitutions?.[param.value];
  return substitution
    ? BasicPrettyPrinter.printExpression(substitution, this.opts)
    : LeafPrinter.param(param);
}
```

Reply:
I think we can't do that if I'm substituting an `identifier`, right?

Reply:
I've reverted these changes